ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,328 @@
+                                    "Iterables should be vectors".to_string(),
+                                ))?,
+                            }
+                        }
+                        IdentifierType::PublicInput(size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                i,
+                            )));
+                        }
+                        IdentifierType::RandomValuesBinding(_, size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                i,
+                            )));
+                        }
+                        _ => Err(SemanticError::InvalidListComprehension(
+                            "Invalid type for a vector".to_string(),
+                        ))?,
+                    }
+                }
+                Iterable::Range(range) => {
+                    return Ok(Expression::Const((range.start() + i) as u64));
+                }
+                Iterable::Slice(ident, range) => {
+                    let ident_type = symbol_table.get_type(ident.name())?;
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                range.start() + i,
+                                CURRENT_ROW,
+                            )));
+                        }
+                        IdentifierType::IntegrityVariable(var_type) => {
+                            match var_type.value() {
+                                VariableType::Vector(vector) => {
+                                    validate_access(i, vector.len())?;
+                                    return Ok(vector[range.start() + i].clone());
+                                }
+                                // TODO: Handle matrix access
+                                _ => Err(SemanticError::InvalidListComprehension(
+                                    "Iterables should be vectors".to_string(),
+                                ))?,
+                            }
+                        }
+                        IdentifierType::PublicInput(size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                range.start() + i,
+                            )));
+                        }
```

Comment:
Public inputs should not be accessible in integrity constraints. Is this checked somewhere later?